src/service/contact_service.rs

```diff
@@ -0,0 +1,196 @@
+    /// exists.
+    async fn add_peer_identity(&self, name: &str, peer_id: &str) -> Result<IdentityPublicData>;
+}
+
+/// The contact service is responsible for managing the contacts and syncing them with the
+/// dht data.
+#[derive(Clone)]
+pub struct ContactService {
+    client: Client,
+    store: Arc<dyn ContactStoreApi>,
+}
+
+impl ContactService {
+    pub fn new(client: Client, store: Arc<dyn ContactStoreApi>) -> Self {
+        Self { client, store }
+    }
+}
+
+#[async_trait]
+impl ContactServiceApi for ContactService {
+    async fn get_contacts(&self) -> Result<Vec<Contact>> {
+        let identities = self.store.get_map().await?;
+        Ok(as_contacts(identities))
+    }
+
+    async fn get_identity_by_name(&self, name: &str) -> Result<IdentityPublicData> {
+        if let Some(identity) = self.store.by_name(name).await? {
+            let public = self
+                .client
+                .clone()
+                .get_identity_public_data_from_dht(identity.peer_id.clone())
+                .await;
+
+            if !public.name.is_empty() && public.ne(&identity) {
+                self.update_identity(name, public.to_owned()).await?;
+                Ok(public)
+            } else {
+                Ok(identity)
+            }
+        } else {
+            Ok(IdentityPublicData::new_empty())
+        }
+    }
+
+    async fn create_identity(
+        &self,
+        name: &str,
+        identity: IdentityPublicData,
+    ) -> Result<IdentityPublicData> {
+        self.store.insert(name, identity.to_owned()).await?;
+        Ok(identity)
+    }
+
+    async fn update_identity(&self, name: &str, identity: IdentityPublicData) -> Result<()> {
+        self.store.update(name, identity).await?;
+        Ok(())
+    }
+
+    async fn delete_identity_by_name(&self, name: &str) -> Result<()> {
+        self.store.delete(name).await?;
+        Ok(())
+    }
+
+    async fn update_identity_name(&self, old_name: &str, new_name: &str) -> Result<()> {
+        self.store.update_name(old_name, new_name).await?;
+        Ok(())
+    }
+
+    async fn add_peer_identity(&self, name: &str, peer_id: &str) -> Result<IdentityPublicData> {
+        let default = IdentityPublicData::new_only_peer_id(peer_id.to_owned());
+        let public = self
+            .client
+            .clone()
+            .get_identity_public_data_from_dht(peer_id.to_owned())
+            .await;
+
+        if public.name.is_empty() {
+            self.store.insert(name, default.clone()).await?;
+            Ok(default)
+        } else {
+            self.store.insert(name, public.clone()).await?;
+            Ok(public)
+        }
+    }
+}
+
+#[derive(Serialize)]
+#[serde(crate = "rocket::serde")]
+pub struct Contact {
+    pub name: String,
+    pub peer_id: String,
+}
+
+// converts identity data to contact data
+fn as_contacts(identities: HashMap<String, IdentityPublicData>) -> Vec<Contact> {
+    let mut contacts_vec: Vec<Contact> = Vec::new();
```

Comment:
Just a matter of preference, but this could be iterator-based in a nice way:

```rust
    identities
        .into_iter()
        .map(|(name, public_data)| Contact {
            name,
            peer_id: public_data.peer_id,
        })
        .collect()
```

I don't think it makes any meaningful difference (in theory it may have less allocations for a large list of contacts), but in my view it reads a bit better in terms of intent.